presto-main/src/main/java/com/facebook/presto/operator/repartition/OptimizedPartitionedOutputOperator.java

```diff
@@ -367,7 +367,8 @@ public OperatorFactory duplicate()
         private final Closer blockLeaseCloser = Closer.create();
 
         // The ArrayAllocator for the buffers used in repartitioning, e.g. PartitionBuffer#serializedRowSizes, BlockEncodingBuffer#mappedPositions.
-        private final ArrayAllocator bufferAllocator = new SimpleArrayAllocator(10_000);
+        // Assuming there are 1000 columns, and each of them uses 3 buffers, and the number of partitions is 333, then we need 1000 * 3 * 333 about 1M buffers.
+        private final ArrayAllocator bufferAllocator = new SimpleArrayAllocator(1_000_000);
```

Comment:
This number is highly dependent on the environment and won't work for all setups. Let's make this limit configurable.